himbaechel/uarch/gowin/pack.cc

```diff
@@ -0,0 +1,1373 @@
+        IdString name_id = ctx->id(name);
+
+        auto dummy_ci = std::make_unique<CellInfo>(ctx, name_id, id_ALU);
+        dummy_ci->params[id_ALU_MODE] = std::string("C2L");
+        return dummy_ci;
+    }
+
+    // create ALU chain
+    void pack_alus(void)
+    {
+        const CellTypePort cell_alu_cout = CellTypePort(id_ALU, id_COUT);
+        const CellTypePort cell_alu_cin = CellTypePort(id_ALU, id_CIN);
+        std::vector<std::unique_ptr<CellInfo>> new_cells;
+
+        log_info("Pack ALUs...\n");
+        for (auto &cell : ctx->cells) {
+            auto ci = cell.second.get();
+            if (ci->cluster != ClusterId()) {
+                continue;
+            }
+            if (is_alu(ci)) {
+                // The ALU head is when the input carry is not a dedicated wire from the previous ALU
+                NetInfo *cin_net = ci->getPort(id_CIN);
+                if (!cin_net || !cin_net->driver.cell) {
+                    log_error("CIN disconnected at ALU:%s\n", ctx->nameOf(ci));
+                }
+                if (CellTypePort(cin_net->driver) != cell_alu_cout || cin_net->users.entries() > 1) {
+                    if (ctx->debug) {
+                        log_info("ALU head found %s. CIN net is %s\n", ctx->nameOf(ci), ctx->nameOf(cin_net));
+                    }
+                    // always prepend first ALU with carry generator block
+                    // three cases: CIN == 0, CIN == 1 and CIN == ?
+                    new_cells.push_back(std::move(alu_add_cin_block(ctx, ci, cin_net)));
+                    CellInfo *cin_block_ci = new_cells.back().get();
+                    // CIN block is the cluster root and is always placed in ALU0
+                    // This is a possible place for further optimization
+                    cin_block_ci->cluster = cin_block_ci->name;
+                    cin_block_ci->constr_z = BelZ::ALU0_Z;
+                    cin_block_ci->constr_abs_z = true;
+
+                    int alu_chain_len = 1;
+                    while (true) {
+                        // add to cluster
+                        if (ctx->debug) {
+                            log_info("Add ALU to the chain (len:%d): %s\n", alu_chain_len, ctx->nameOf(ci));
+                        }
+                        cin_block_ci->constr_children.push_back(ci);
+                        NPNR_ASSERT(ci->cluster == ClusterId());
+                        ci->cluster = cin_block_ci->name;
+                        ci->constr_abs_z = false;
+                        ci->constr_x = alu_chain_len / 6;
+                        ci->constr_y = 0;
+                        ci->constr_z = alu_chain_len % 6;
+                        // XXX I2 is pin C which must be set to 1 for all ALU modes except MUL
+                        // we use only mode 2 ADDSUB so create and connect this pin
+                        ci->addInput(id_I2);
+                        ci->connectPort(id_I2, ctx->nets[ctx->id("$PACKER_VCC")].get());
```

Comment:
nitpick: ` ctx->nets.at(ctx->id("$PACKER_VCC")).get()` rather than `[]` - this way it will fail if for some reason the net doesn't exist rather than silently creating a `nullptr` net and then failing in a harder to debug way